packages/rum-core/src/domain/foregroundContexts.ts

```diff
@@ -0,0 +1,123 @@
+import {
+  ServerDuration,
+  Configuration,
+  noop,
+  addEventListener,
+  DOM_EVENT,
+  RelativeTime,
+  elapsed,
+  relativeNow,
+  toDuration,
+  toServerDuration,
+} from '@datadog/browser-core'
+import { ForegroundContext, InForegroundPeriod } from '../rawRumEvent.types'
+
+const MAX_NUMBER_OF_FOCUSED_TIME = 500
+
+export interface ForegroundContexts {
+  getInForeground: (startTime: RelativeTime) => ForegroundContext | undefined
+  getInForegroundPeriods: (startTime: RelativeTime, duration: ServerDuration) => InForegroundPeriod[] | undefined
+  stop: () => void
+}
+
+export interface FocusPeriod {
+  start: RelativeTime
+  end?: RelativeTime
+}
+
+let focusPeriods: FocusPeriod[] = []
+
+export function startForegroundContexts(configuration: Configuration) {
+  if (!configuration.isEnabled('track-foreground')) {
+    return {
+      getInForeground: () => undefined,
+      getInForegroundPeriods: () => undefined,
+      stop: noop,
+    }
+  }
+  if (document.hasFocus()) {
+    addNewFocusPeriod()
+  }
+
+  const { stop: stopFocusTracking } = trackFocus(addNewFocusPeriod)
+  const { stop: stopBlurTracking } = trackBlur(closeFocusPeriod)
+  return {
+    getInForeground,
+    getInForegroundPeriods,
+    stop: () => {
+      focusPeriods = []
+      stopFocusTracking()
+      stopBlurTracking()
+    },
+  }
+}
+
+function addNewFocusPeriod() {
+  const now = relativeNow()
+  if (focusPeriods.length > MAX_NUMBER_OF_FOCUSED_TIME) {
+    return
+  }
+  const lastIndex = focusPeriods.length - 1
+  if (lastIndex >= 0 && focusPeriods[lastIndex].end == null) {
+    // already here
+    if (focusPeriods[lastIndex].start === now) {
```

Comment:
Yes 
![image](https://user-images.githubusercontent.com/44997281/118842450-1f676200-b8c1-11eb-83df-2190bd849307.png)
https://app-491fd269e5f78f447a9037f3b324e786.datadoghq.com/rum/explorer?event=AQAAAXmEo5Ij1bg8SgAAAABBWG1FbzVIeU1UYXN0Z01mc2FYZ3FR&from_ts=1621421066948&live=true&query=%40usr.email%3Athibaut.gery%40datadoghq.com%20%40browser.name%3AChrome%20%40session.has_replay%3Atrue%20%40view.id%3A36936bd0-b817-475d-be01-eff2acf4c053&tab=view&to_ts=1621435466948&p_tab=attributes

Reply:
Did you understand what happen?
if yes, it could nice to add a test explaining this case
if no, I'd be in favor of trying to

Reply:
> Did you understand what happen?

Not really, I've added a bunch of internalMonitoring to debug this

>  if yes, it could nice to add a test explaining this case

I've got test case but it's just very artificial and I do not get the exact root cause